### DISCUSSION TITLE
Correct Configuration Methods for Pipelines when Using Dagit

### DISCUSSION THREAD

**Discussion by prinapatel** [dagster-io/dagster#3630] [Q&A]
Hi there,

I have recently started using Dagster and I'm confused about how the various modes of configuration work for solids. I've been using the following toy example code to investigate how it all fits together: 

```
from dagster import InputDefinition, OutputDefinition, pipeline, solid, execute_pipeline, Nothing, List, String, Field

#@solid(config_schema={
#            "filename": Field(String,is_required=False,default_value="hello_world_file",description="Name of the file")
#        }
#)
@solid()
def func0(context,filename) -> Nothing:
    context.log.info("has this worked? filename is {file}".format(file=context.solid_config['filename']))

@solid(input_defs=[InputDefinition("start", Nothing)],
       output_defs=[OutputDefinition(name='func1output',dagster_type=List[String])])
def func1(context) -> List:
    context.log.info("Do i see anything here?")
    return ['hello','world']

@solid(input_defs=[InputDefinition(name='func2input',dagster_type=List[String])])
def func2(context,func2input) -> Nothing:
    for x in func2input:print(x)

@pipeline
def my_pipeline():
    func2(func1(func0()))


#run_config={
#    "solids": {
#        "func0": {"inputs": {"filename": {"value": "test.json"}}}
#    }
#}
```

What I'm trying to achieve is having a solid (`func0` in this case) that takes an input variable (`filename`) and for that variable to default to some pre-configured value that can then be changed in dagit if the user wishes. Ideally, I want to put this in the `@solid()` decorator as I probably will have many functions with many parameters, I also don't want to have to write them into the dagit GUI if I can avoid it, as most of the time the default values should be fine to start with. I've read through the docs on this but wasn't able to run most of the examples so am not sure if I am missing something here. I couldn't understand from the docs if all the methods would work with all methods of execution so I had to resort to trial-and-error to try to understand this.  

I am running dagtser/dagit 0.10.0.

1. In the first case, where I just use the `@solid()` decorator and then run this script with dagit, in the Playground I get the error saying there is config missing which is as I expect. 

(
As an aside, when I press the Scaffold Missing Config button what I get is:
```
solids:
  func0:
    inputs:
      filename: <selector>
```
with a red cross at line 4 (indicating that the value for the selector `filename` should be a dict).
)

2. When I then try using the `@solid(config_schema = ...)` that has the `filename` configured with a `Field` that has a `default_value`, upon running in dagit I still get an error saying that I need a configuration as above. Even when I populate the `filename` section in dagit and then execute the pipeline it ignores the values I entered in the Playground and then uses the `config_schema` `default_value`. It's as if the Playground does not see the `config_schema`? 

3. Similarly, when I uncomment the `run_config` this also seems to be ignored in the Playground and on execution the config I type in the playground is also ignored in favour of the `default_value` that is in the `config_schema` for that solid. 

So basically what I find is that I cannot execute the `my_pipeline` without writing something into the Playground manually, but then when I do that it still uses the configuration that is coded into the `solid` decorator, so the typing is necessary but ignored. 

Can anyone help with what is the correct way to write pipelines to be pre-configured for running with dagit. Note that I also found that with the above code which runs, if I add a `main` function and then try to run from the dagster CLI I get errors which also seems unexpected. 

Thanks in advance for any help!

**Comment by sryza** [comment]
Hi @prinapatel, there should be two ways to get this to work, depending on whether you want to model the filename as an input or as part of the solid config. Let me know if these don't work for you.

### Solid config
```
@solid(config_schema={
            "filename": Field(String,is_required=False,default_value="hello_world_file",description="Name of the file")
        }
)
def func0(context) -> Nothing:
    context.log.info("has this worked? filename is {file}".format(file=context.solid_config['filename']))
```

Then, in playground, if you want to override:
```
solids:
  func0:
    config:
      filename: the_name_of_your_file
```

### Input

```
@solid(input_defs=[InputDefinition("filename", default_value="hello_world_file")])
def func0(context,filename) -> Nothing:
    context.log.info("has this worked? filename is {file}".format(file=context.solid_config['filename']))
```

Then, in playground, if you want to override:
```
solids:
  func0:
    inputs:
      filename:
        value: the_name_of_your_file
```

**Reply by prinapatel** [reply]
Hi @sryza,

Thanks for the response. 

When I try the first suggestion, the Playground config doesn't override the default value in the solid config: i.e. the output is "has this worked? filename is moring Prina".

The second suggestion does execute because `context.solid_config['filename']` doesn't exist. I've attached the full error if that is useful at the bottom.

So in both cases, even with the config done in the playground to override the configs set in the code the execution is not using it? 

Cheers,

Prina

![Screenshot 2021-01-29 at 16 51 46](https://user-images.githubusercontent.com/6458085/106303921-9fa28280-6252-11eb-8325-e7eb2e323ff0.png)

**Reply by sryza** [reply]
Sorry, the second example should look like this:
```
@solid(input_defs=[InputDefinition("filename", default_value="hello_world_file")])
def func0(context,filename) -> Nothing:
    context.log.info("has this worked? filename is {file}".format(file=filename))
```

**Reply by sryza** [reply]
> the output is "has this worked? filename is moring Prina".

where are you including the text "moring Prina"?

**Reply by prinapatel** [reply]
"moring Prina" is the default in solid decorator:
```
@solid(config_schema={
            "filename": Field(String,is_required=False,default_value="moring Prina",description="Name of the file")
        }
)
```

So this is not getting overridden by the playground configuration.

**Reply by prinapatel** [reply]
> Sorry, the second example should look like this:
> 
> ```
> @solid(input_defs=[InputDefinition("filename", default_value="hello_world_file")])
> def func0(context,filename) -> Nothing:
>     context.log.info("has this worked? filename is {file}".format(file=filename))
> ```

Sorry - once I change it to this I see now that the override works if set in the Playground otherwise it defaults to the value set in the `InputDefinition` for `filename`.

Is this the expected behaviour? I would have thought that both `config_schema` and `input_defs` should give the same result?

**Reply by sryza** [reply]
Weird - I just tried out the config method on my local machine and it worked as expected.  I.e. the config that I included in dagit took effect.  What version are you running?  What happens if you omit the default value for the config Field, but leave everything else the same?

**Reply by prinapatel** [reply]
Hey @sryza,

That is weird. I am also running locally. It's also interesting that if I use the set the `config_schema` in the `solid()` then I cannot execute the pipeline as the Playground gives me an `Missing required config entry "solids" at the root.` error. So I have to put a config in there and then I have to input a value for the `filename`. I tested it by inputting something different from the `default_value` in the `config_scheama` which is how I can tell that the Playground configuration is being ignored. Is this expected behaviour?

With the `input_defs` I don't get the `Missing required config entry "solids" at the root.` so I can execute the pipeline and it picks up the `default_value` that is set. This is what I'd expect. 

If I change the decorator to:
```
@solid(config_schema={
            "filename": Field(String,description="Name of the file")
        }
)
```
I still get the `Missing required config entry "solids" at the root.` error in the Playground, so I still need to input a configuration, but this now also looks different:
```
solids:
  func0:
    config:
      filename: ""
    inputs:
      filename: <selector>
```

when I configure that to:
```
solids:
  func0:
    config:
      filename: "test.json"
    inputs:
      filename: 
        value: "test2.json"
```
and execute, the output reflects the `inputs.filename.value` `"test2.json"`. 

I'm running version 0.10.0. 

Thanks Prina

**Reply by sryza** [reply]
Hey @prinapatel - it looks like you're trying out the config approach and the input approach at the same time.  Have you tried the config approach without specifying any inputs?
